pkg/compactor/bucket-index-block-ids-fetcher.go

```diff
@@ -0,0 +1,65 @@
+package compactor
+
+import (
+	"context"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	"github.com/cortexproject/cortex/pkg/storage/tsdb/bucketindex"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/objstore"
+)
+
+type BucketIndexBlockIDsFetcher struct {
+	logger      log.Logger
+	bkt         objstore.Bucket
+	userID      string
+	cfgProvider bucket.TenantConfigProvider
+}
+
+func NewBucketIndexBlockIDsFetcher(logger log.Logger, bkt objstore.Bucket, userID string, cfgProvider bucket.TenantConfigProvider) *BucketIndexBlockIDsFetcher {
+	return &BucketIndexBlockIDsFetcher{
+		logger:      logger,
+		bkt:         bkt,
+		userID:      userID,
+		cfgProvider: cfgProvider,
+	}
+}
+
+func (f *BucketIndexBlockIDsFetcher) GetActiveAndPartialBlockIDs(ctx context.Context, ch chan<- ulid.ULID) (partialBlocks map[ulid.ULID]bool, err error) {
+	// Fetch the bucket index.
+	idx, err := bucketindex.ReadIndex(ctx, f.bkt, f.userID, f.cfgProvider, f.logger)
+	if errors.Is(err, bucketindex.ErrIndexNotFound) {
+		// This is a legit case happening when the first blocks of a tenant have recently been uploaded by ingesters
+		// and their bucket index has not been created yet.
```

Comment:
Can we fallback to use base ID fetcher in this case? Imagine users migrate from non bucket index to bucket index enabled, they could have a lot of blocks but no bucket index

Reply:
Added fallback to the BaseBlockIDsFetcher for this case.